### PR TITLE
Directory move fix

### PR DESCRIPTION
**Pull request by Jlalond**
Fix for #40013 

Allowing the moving of directories with different casing.

Thus if in a case-sensitive file system renaming Directory .\foo to .\FOO should still be allowed.

But still preventing a directory having .\foo .\fOO .\FOO

As I am no expert on inter-opting with the OS I look forward to feedback

(And I went out on a bit of a limb with the design of this change)

### CODE REVIEW

**Comment by Jlalond** [comment]
@JeremyKuhne after digging through my logs, it looks like it's failing because when I expect the interop to return me `FILE_ALREADY_EXISTS` it silently continues or no-ops (Hard to tell through the lenses of a unit test). Do we want to build something platform specific for these *NIX builds, or is this potentially an issue with the interop code?

**Comment by Jlalond** [comment]
@JeremyKuhne / @danmosemsft could I get a review on this?

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 1]
```diff
@@ -268,23 +268,38 @@ public static void Move(string sourceDirName, string destDirName)
...
             string destPath = PathInternal.EnsureTrailingSeparator(fulldestDirName);
 
-            StringComparison pathComparison = PathInternal.StringComparison;
+            // Here we have a bool to check if they're equal, but with different casing
+            // We allowing moving foo to the same directory, but renaming with a different case.
+            bool sameDirectoryIgnoreCase =
+                string.Equals(sourcePath, destPath, StringComparison.OrdinalIgnoreCase);
 
-            if (string.Equals(sourcePath, destPath, pathComparison))
+            StringComparison pathComparison = StringComparison.Ordinal;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: There is no need for a local for this.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 2]
```diff
@@ -268,23 +268,38 @@ public static void Move(string sourceDirName, string destDirName)
...
+
+            // If the paths are the exact same, fail.
+            if (sameDirectoryIgnoreCase
+                && string.Equals(sourcePath, destPath, pathComparison))
+            {
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: The rest of this method does not use brackets for the throw statements.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 3]
```diff
@@ -268,23 +268,38 @@ public static void Move(string sourceDirName, string destDirName)
...
                 throw new IOException(SR.IO_SourceDestMustBeDifferent);
+            }
+
 
             string sourceRoot = Path.GetPathRoot(sourcePath);
             string destinationRoot = Path.GetPathRoot(destPath);
             if (!string.Equals(sourceRoot, destinationRoot, pathComparison))
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: You should add tests for roots that differ by case. Moving `C:\foo` to `c:\bar` should not fail and presumably it does now.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 4]
```diff
@@ -268,23 +268,38 @@ public static void Move(string sourceDirName, string destDirName)
             string fulldestDirName = Path.GetFullPath(destDirName);
             string destPath = PathInternal.EnsureTrailingSeparator(fulldestDirName);
 
-            StringComparison pathComparison = PathInternal.StringComparison;
+            // Here we have a bool to check if they're equal, but with different casing
+            // We allowing moving foo to the same directory, but renaming with a different case.
+            bool sameDirectoryIgnoreCase =
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: For clarity I would move this and the comment inline with the check where we throw for the existing directory. I'd keep the comment simpler too- "We allow moving to the same name with different casing."

File: src/System.IO.FileSystem/tests/Directory/Move.cs [hunk 5]
```diff
@@ -212,6 +212,57 @@ public void Path_Longer_Than_MaxLongPath_Throws_Exception()
...
 
+        [Fact]
+        public void ThrowIOExceptionWhenDirectoryExists()
+        {
+            Directory.CreateDirectory($"{TestDirectory}/foo");
+            Assert.Throws<IOException>(() => Directory.Move($"{TestDirectory}/foo", $"{TestDirectory}/foo"));
+        }
+
+        [Fact]
+        public void MoveDirectoryToNewDirectoryButWithDifferentCasing()
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/Move.cs]: nit: the rest of the test methods you're using underscore for the name

File: src/System.IO.FileSystem/tests/Directory/Move.cs [hunk 6]
```diff
@@ -212,6 +212,57 @@ public void Path_Longer_Than_MaxLongPath_Throws_Exception()
...
+        {
+            Directory.CreateDirectory(Path.Combine(TestDirectory, "foo"));
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/Move.cs]: I would add tests that don't match the actual file system casing. What happens when you have `FOO` and you try to move `foo` to `FOO`?

**Comment by JeremyKuhne** [comment]
cc: @carlossanlop

**Comment by Jlalond** [comment]
@carlossanlop / @JeremyKuhne can I get a re-review? Crytography failed for UWP, so if you have any review comments I can push again and it'll rerun, and I suspect it'll clear.

**Comment by Jlalond** [comment]
@JeremyKuhne / @carlossanlop I implemented the changes @JeremyKuhne suggested (To the best of my understanding) OSX timed out, but all the builds otherwise are passing. Could I get a re-review, and if anyone can re-run that task I'd appreciate it. As I cannot without re-running all tests for the language.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 7]
```diff
@@ -264,31 +264,67 @@ public static void Move(string sourceDirName, string destDirName)
...
 
             string fulldestDirName = Path.GetFullPath(destDirName);
             string destPath = PathInternal.EnsureTrailingSeparator(fulldestDirName);
+            string destDirectoryName = Path.GetFileName(fulldestDirName);
 
-            StringComparison pathComparison = PathInternal.StringComparison;
+            bool sameDirectoryIgnoreCase =
+                string.Equals(sourcePath, destPath, StringComparison.OrdinalIgnoreCase);
+            bool sameDirectoryNameIgnoreCase
+                = !string.IsNullOrEmpty(sourceDirectoryName) && string.Equals(sourceDirectoryName, destDirectoryName, StringComparison.OrdinalIgnoreCase);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: There is no need for `IsNullOrEmpty`, `string.Equals` handles that. Additionally, don't do this here, put it inline with the condition below. It isn't used until then which makes this really hard to follow.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 8]
```diff
@@ -264,31 +264,67 @@ public static void Move(string sourceDirName, string destDirName)
...
+                !string.Equals(sourceRoot, destinationRoot, StringComparison.OrdinalIgnoreCase))
                 throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
 
             // Windows will throw if the source file/directory doesn't exist, we preemptively check
             // to make sure our cross platform behavior matches NetFX behavior.
             if (!FileSystem.DirectoryExists(fullsourceDirName) && !FileSystem.FileExists(fullsourceDirName))
                 throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
 
-            if (FileSystem.DirectoryExists(fulldestDirName))
+            if (!sameDirectoryIgnoreCase // This check is to allowing renaming of directories
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: This needs a better comment and should be ahead of the check. Perhaps something like:

``` C#
\\ Don't allow moving to an existing directory unless only the casing is different (to allow renames)
```

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 9]
```diff
@@ -264,31 +264,67 @@ public static void Move(string sourceDirName, string destDirName)
...
+                && FileSystem.DirectoryExists(fulldestDirName))
+                throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
+
+            if (!sameDirectoryIgnoreCase
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: This is really complicated to have no comment. What exactly are you looking for? Why are you looking for the same directory with a different casing?

File: src/System.IO.FileSystem/src/System/IO/Directory.cs (hunk 7 above)
Comment (Jlalond) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Fair. If you don't mind clarifying I'm doing this as I want to check if a directory with a case-sensitive version of that already exists in a different directory. So if I try to move `foo` to a directory with `FOO`, it wouldn't fail when I wanted it to. Do we only want to supper move as a renaming function? If so I can seriously reduce this complexity and make the contains not have lots of heap allocations for the strings.

Comment (JeremyKuhne) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Why doesn't `FileSystem.DirectoryExists(fulldestDirName)` do that for you? If you're moving `/A/B` to `/B/b` when there is a `/B/B` or  `C:\A\B` to `C:\B\b` when there is a `C:\B\B` that would succeed/fail appropriately (given normal casing rules for Unix/Windows).

Comment (Jlalond) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: This is an implementation question by me. Directory.Exists doesn't seem to check if a directory of a different case exists there. When I was debugging on W10, I moved `foo` into a folder with `fOo` . Maybe I'm misremembering but do we just want to let the OS see if the directory with the same case? If so that's a lot easier than my current scanning of all directories.

Comment (JeremyKuhne) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Letting the OS check whether casing is a conflict is the best route to go. If the OS tells us it doesn't exist, then there is no conflict and we should let it through.

**Comment by Jlalond** [comment]
Hey @JeremyKuhne, Implemented your suggestions. I actually got really tired one night and commited moronic code, but serendipitously I think I found some odd behavior. On all our Linux builds moving `foo` into a directory with `FOO` works, _except Debug builds for x64 and Release for Redhat_. I found that pretty funky, is that expected behavior? (Also I should be good for another PR to destroy my confidence whenever you're ready :D )

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 10]
```diff
@@ -268,22 +268,33 @@ public static void Move(string sourceDirName, string destDirName)
...
             // to make sure our cross platform behavior matches NetFX behavior.
             if (!FileSystem.DirectoryExists(fullsourceDirName) && !FileSystem.FileExists(fullsourceDirName))
                 throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
 
-            if (FileSystem.DirectoryExists(fulldestDirName))
+            if (!sameDirectoryIgnoreCase // This check is to allowing renaming of directories
+                && FileSystem.DirectoryExists(fulldestDirName))
+                throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
+
+            // If the direcotires aren't the same and the OS says the directory exists already, fail.
```
Comment (carlossanlop) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]:
```suggestion
            // If the directories aren't the same and the OS says the directory exists already, fail.
```

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 11]
```diff
@@ -268,22 +268,33 @@ public static void Move(string sourceDirName, string destDirName)
...
+            if (!sameDirectoryIgnoreCase && Directory.Exists(fulldestDirName))
                 throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
 
             FileSystem.MoveDirectory(fullsourceDirName, fulldestDirName);
```
Comment (carlossanlop) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: Let's assume you have a folder "C:\FOO\BAR"

What would the result be is if you call the method with these changes, using the following two arguments?:

> `sourceDirName = "C:\\foo\\BAR"`
> `destDirName = "C:\\FOO\\BAR"`

From what I understand:

1. Line 271: `sameDirectoryIgnoreCase` is `true`. That's ok.
2. Line 275: `sameDirectoryIgnoreCase` is `true` but the ordinal comparison is `false`, so we won't throw here. Should we have thrown at this point? The leaf directory was the exact same, but the parent was not.
3. Line 283: `!sameDirectoryIgnoreCase` evaluates to `false`, so we will not do the second check where we should verify if the roots are in the same volume. This verification should not be skipped because Windows does not allow moving between volumes.
4. Line 289: The source FileSystem directory exists and the source FileSystem file exists, so we won't throw. That's ok.
5. Line 292: Again, `!sameDirectoryIgnoreCase` evaluates to `false`, so we will not do the second check where we should verify if the destination FileSystem directory exists, even though we should.
6. Line 297: `!sameDirectoryIgnoreCase` evaluates to `false`, so we don't check if the destination Directory exists, even though we should.
7. Steps 5 and 6 made me wonder if we should be evaluating both `FileSystem.Exists` and `Directory.Exists` in the same `if` instead of separately.

We would end up reaching line 300, and I am not sure what the expected result would be.

cc @JeremyKuhne

Comment (Jlalond) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: @carlossanlop as @JeremyKuhne  hasn't responded. I should check if the paths are ordinally (or whatever the OS says) the same, other than directory names, if the paths are the same, but the names are case-variants of themselves. I should continue, otherwise fail for question 2.

Do you agree? This PR is kind of in limbo now and I'd like to get it closed out

File: src/System.IO.FileSystem/tests/Directory/Move.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: Is this addressed?

File: src/System.IO.FileSystem/src/System/IO/Directory.cs (hunk 11 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Sorry @Jlalond! Been distracted with WinForms work. Your answer seems plausible to me. @carlossanlop can you sanity check this?

I had a test question that I've responded to as well.

**Comment by Jlalond** [comment]
@carlossanlop @JeremyKuhne I switched over to checking if the dir's were case variants, and deferring to the OS for the paths. Could I get a review, and sorry this PR seems to be going on forever, but this was much more nuanced than I originally realized.

File: src/System.IO.FileSystem/tests/Directory/Move.cs [hunk 12]
```diff
@@ -212,10 +212,93 @@ public void Path_Longer_Than_MaxLongPath_Throws_Exception()
             });
         }
 
+        [Fact]
+        public void MoveDirectory_ThrowIOExceptionWhenDirectoryExists()
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/Directory/Move.cs]: This is testing something more specific? "when directory exists and destination is the same" .. those seem like 2 cases?

Comment (Jlalond) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: Agreed, something along the lines of `ThrowIOExceptionWhenMovingToADirectoryThatAlreadyExists`? Incredibly wordy, but I running low on ideas on how to shorten it

Comment (danmosemsft) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: I meant, this test is testing two things at once  -- moving a directory which exists to a different directory that already exists and also moving a directory which exists to itself. Seems like those ought to be different tests, as typicaly the implementation would check for "destination exists" and "destination same as source" separately in the code. (I didn't look)

Comment (Jlalond) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: Ha, @danmosemsft I was just in the middle of replying that I misread your comment. I should absolutely split this into two test cases with better names.

File: src/System.IO.FileSystem/tests/Directory/Move.cs (hunk 6 above)
Comment (Jlalond) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: @JeremyKuhne just so I understand, you mean I have a directory of `FOO` and I ask to move a directory `foo` to `FOO`. So sending a case-variant as an argument into Move?

**Comment by Jlalond** [comment]
@carlossanlop Does my rework make sense to you? At what point should I just give up 🤣

File: src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs [hunk 13]
```diff
@@ -336,6 +336,9 @@ public static void MoveDirectory(string sourceFullPath, string destFullPath)
                 if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
                     throw Win32Marshal.GetExceptionForWin32Error(Interop.Errors.ERROR_PATH_NOT_FOUND, sourceFullPath);
 
+                if (errorCode == Interop.Errors.ERROR_ALREADY_EXISTS)
```
Comment (carlossanlop) [inline src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs]: If this new exception case gets

File: src/System.IO.FileSystem/src/System/IO/Directory.cs (hunk 11 above)
Comment (carlossanlop) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: The changes are more clear now. Thanks for addressing. I see you also added a unit test to cover this example.

File: src/System.IO.FileSystem/tests/Directory/Move.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: @Jlalond yes "FOO" exists as a directory and you try to move "foo" (which also exists) to "FOO". (Sorry about the slow response.) 🙏

Comment (Jlalond) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: I believe I addressed this starting [299](https://github.com/dotnet/corefx/pull/40570/files#diff-4b86f8ca99b5dc228df21c6e6e417034R299) can you confirm my logic? Specifically around the case-sensitive vs insensitive OS

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

File: src/System.IO.FileSystem/tests/Directory/Move.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/tests/Directory/Move.cs]: I think we're ok here.
